WordPress/gutenberg
# Add experimental API for registering block bindings on the editor side

🧪 **Experimental prototype code ahead!** This PR is meant more as an anchoring point in the ongoing discussion about Block Bindings rather than a concrete proposal.

## What?
Refactors the editor-side code to add an interface through which new custom sources can be registered.

## Why?
Following the discussion in https://github.com/WordPress/gutenberg/issues/53300 it makes sense to have a clear API that allows registration of custom sources. This registration should be separate from the "core" which handles the block bindings logic. Each and every source (including custom fields!) should use that same interface in the future.

## How?
- New `registerBlockBinding({ name, edit })` function.
- New `useAvailableBlockAttributes(blockName)` React hook. This hook should be used in the `edit` component passed to `registerBlockBindings()`. Given block name as input, it returns the attributes allowed to be connected to custom sources.

## Testing Instructions
1. Enable the "block connections" Gutenberg experiment (side note: this should be renamed to Block bindings experiement)
2. Open a post or page.
3. Insert a paragraph block.
4. Make sure that the custom fields work as expected.
